tsdb/head.go

```diff
@@ -1825,7 +2075,11 @@ func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkC
 	c := s.head()
 
 	if c == nil {
-		c = s.cut(t)
+		// Out of order sample. Don't cut a chunk with wrong timestamp.
+		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
+			return false, chunkCreated
+		}
+		c = s.cut(t, chunkDiskMapper)
```

Comment:
I would say splitting is going to make it more complex to understand. The way you can look at it is:
1. `cut()` is called to create a new chunk. We are checking for `nil` here because if this was the first sample being appended, then we want to create a new chunk before we proceed.
2. The non-nil chunk inside `cut()` is because we don't want to mmap the in-memory chunk. If we are going to create a new chunk and if that is going to be in-memory chunk, we don't mmap it. If there was already a chunk (`s.head() != nil`), then we m-map it.

Comparing this nil check with the nil check inside `cut()` are not much related as their purpose is different. I will add comments.

Reply:
What about something like this?


```
func (s *memSeries) cut(mint int64, chunkDiskMapper *chunks.ChunkDiskMapper) *memChunk {
	if s.headChunk != nil {
		// There is already a chunk, m-map it before we replace it with a new one.
		chunkRef, err := chunkDiskMapper.WriteChunk(s.ref, s.headChunk.minTime, s.headChunk.maxTime, s.headChunk.chunk)
		if err != nil {
			if err == chunks.ErrChunkDiskMapperClosed {
				// The head is being closed, hence chunkDiskMapper is closed.
				// Return the same head chunk as there wont be more samples appended anyway.
				return s.headChunk
			}
			panic(err)
		}
		s.mmappedChunks = append(s.mmappedChunks, &mmappedChunk{
			ref:        chunkRef,
			numSamples: uint16(s.headChunk.chunk.NumSamples()),
			minTime:    s.headChunk.minTime,
			maxTime:    s.headChunk.maxTime,
		})
	}

	return s.newHead(mint)
}

func (s *memSeries) newHead(mint int64) *memChunk {
	s.headChunk = &memChunk{
		chunk:   chunkenc.NewXORChunk(),
		minTime: mint,
		maxTime: math.MinInt64,
	}

	// Set upper bound on when the next chunk must be started. An earlier timestamp
	// may be chosen dynamically at a later point.
	s.nextAt = rangeForTimestamp(mint, s.chunkRange)

	app, err := s.headChunk.chunk.Appender()
	if err != nil {
		panic(err)
	}
	s.app = app
	return s.headChunk
}
```

then this part of the code will simple become as 
```
if c == nil {
		// Out of order sample. Don't cut a chunk with wrong timestamp.
		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
			return false, chunkCreated
		}
		// There is no chunk in this series yet, create the first chunk.
		c = s.newHead(t)
		chunkCreated = true
	}
```

Reply:
I still very much prefer to have a single method to create a new chunk, makes it much simpler to call it anywhere and no need to think if we have a head or no before calling `cut`. I am still not convinced that the check for head chunk being nil is a problem here. Like some methods have some small tasks to perform at the beginning on a method, similarly here we just mmap a chunk when head chunk already exists. I will move the mmaping part to a different method and call it inside cut if that makes it simpler; to have a single way to create new chunk.

Reply:
yep that works as well, I could follow the code much better, thanks